Userland/Services/WebContent/BackingStoreManager.cpp

```diff
@@ -0,0 +1,143 @@
+}
+
+#ifdef AK_OS_MACOS
+static Optional<Core::MachPort> s_browser_mach_port;
+void BackingStoreManager::set_browser_mach_port(Core::MachPort&& port)
+{
+    s_browser_mach_port = move(port);
+}
+#endif
+
+BackingStoreManager::BackingStoreManager(PageClient& page_client)
+    : m_page_client(page_client)
+{
+    m_backing_store_shrink_timer = Core::Timer::create_single_shot(3000, [this] {
+        resize_backing_stores_if_needed(false);
+    });
+}
+
+void BackingStoreManager::restart_resize_timer()
+{
+    m_backing_store_shrink_timer->restart();
+}
+
+void BackingStoreManager::reallocate_backing_stores(Gfx::IntSize size)
+{
+#ifdef AK_OS_MACOS
+    if (!s_iosurface_usage_disabled) {
+        auto back_iosurface = Core::IOSurfaceHandle::create(size.width(), size.height());
+        auto back_iosurface_port = back_iosurface.create_mach_port();
+
+        auto front_iosurface = Core::IOSurfaceHandle::create(size.width(), size.height());
+        auto front_iosurface_port = front_iosurface.create_mach_port();
+
+        m_front_bitmap_id = m_next_bitmap_id++;
+        m_back_bitmap_id = m_next_bitmap_id++;
+
+        m_front_bitmap = Gfx::Bitmap::create_wrapper(Gfx::BitmapFormat::BGRA8888, size, size.width() * front_iosurface.bytes_per_element(), front_iosurface.data(), [handle = move(front_iosurface)] {}).release_value();
+        m_back_bitmap = Gfx::Bitmap::create_wrapper(Gfx::BitmapFormat::BGRA8888, size, size.width() * back_iosurface.bytes_per_element(), back_iosurface.data(), [handle = move(back_iosurface)] {}).release_value();
+
+        Core::Platform::BackingStoreMetadata metadata;
+        metadata.page_id = m_page_client.m_id;
+        metadata.front_backing_store_id = m_front_bitmap_id;
+        metadata.back_backing_store_id = m_back_bitmap_id;
+
+        Core::Platform::MessageWithBackingStores message;
+
+        message.header.msgh_remote_port = s_browser_mach_port->port();
+        message.header.msgh_local_port = MACH_PORT_NULL;
+        message.header.msgh_bits = MACH_MSGH_BITS(MACH_MSG_TYPE_COPY_SEND, 0) | MACH_MSGH_BITS_COMPLEX;
+        message.header.msgh_size = sizeof(message);
+        message.header.msgh_id = Core::Platform::BACKING_STORE_IOSURFACES_MESSAGE_ID;
+
+        message.body.msgh_descriptor_count = 3;
+
+        message.front_descriptor.name = front_iosurface_port.port();
+        message.front_descriptor.disposition = MACH_MSG_TYPE_COPY_SEND;
+        message.front_descriptor.type = MACH_MSG_PORT_DESCRIPTOR;
+
+        message.back_descriptor.name = back_iosurface_port.port();
+        message.back_descriptor.disposition = MACH_MSG_TYPE_COPY_SEND;
+        message.back_descriptor.type = MACH_MSG_PORT_DESCRIPTOR;
+
+        message.metadata.address = (void*)&metadata;
+        message.metadata.size = sizeof(Core::Platform::BackingStoreMetadata);
+        message.metadata.deallocate = FALSE;
+        message.metadata.copy = MACH_MSG_VIRTUAL_COPY;
+        message.metadata.type = MACH_MSG_OOL_DESCRIPTOR;
```

Comment:
uhhh, I guess this would work, but why do it like that, rather than including the data in the message inline?

Reply:
oh duh, I though I have to use descriptors for everything with Mach ports 😅 changed it to inline data.